packages/eslint-plugin/tests/rules/prefer-at.test.ts

```diff
@@ -0,0 +1,475 @@
+import type { RuleModule } from '@typescript-eslint/utils/dist/ts-eslint';
+
+import rule from '../../src/rules/prefer-at';
+import type { InvalidTestCase, ValidTestCase } from '../RuleTester';
+import { getFixturesRootDir, RuleTester } from '../RuleTester';
+
+const rootDir = getFixturesRootDir();
+
+const ruleTester = new RuleTester({
+  parserOptions: {
+    ecmaVersion: 2018,
+    tsconfigRootDir: rootDir,
+    project: './tsconfig.json',
+  },
+  parser: '@typescript-eslint/parser',
+});
+
+type TMessageIds = typeof rule extends RuleModule<infer T, unknown[]>
+  ? T
+  : never;
+type TOptions = typeof rule extends RuleModule<string, infer T> ? T : never;
+
+interface Declaration {
+  name: string;
+
+  type: string;
+}
+
+const declarations: Array<Declaration> = [
+  {
+    name: 'str',
+    type: 'string',
+  },
+  {
+    name: 'str',
+    // eslint-disable-next-line @typescript-eslint/internal/prefer-ast-types-enum
+    type: 'String',
+  },
+  {
+    name: 'arr',
+    type: 'Array<string>',
+  },
+  {
+    name: 'arr',
+    type: 'Int8Array',
+  },
+  {
+    name: 'arr',
+    type: 'Uint8Array',
+  },
+  {
+    name: 'arr',
+    type: 'Uint8ClampedArray',
+  },
+  {
+    name: 'arr',
+    type: 'Int16Array',
+  },
+  {
+    name: 'arr',
+    type: 'Uint16Array',
+  },
+  {
+    name: 'arr',
+    type: 'Int32Array',
+  },
+  {
+    name: 'arr',
+    type: 'Float32Array',
+  },
+  {
+    name: 'arr',
+    type: 'Uint32Array',
+  },
+  {
+    name: 'arr',
+    type: 'Float64Array',
+  },
+  {
+    name: 'arr',
+    type: 'BigInt64Array',
+  },
+  {
+    name: 'arr',
+    type: 'BigUint64Array',
+  },
+];
+
+const additionalDeclarations: Array<
+  (Declaration & { type: 'number' }) | number
+> = [
+  1,
+  123,
+  {
+    name: 'i',
+    type: 'number',
+  },
+];
+
+interface InvalidCodeWithMetadata {
+  code: string;
+
+  line: number;
+
+  pos: {
+    start: number;
+    end: number;
+  };
+}
+
+interface CodeGenerator {
+  valid: string;
+
+  invalid: InvalidCodeWithMetadata;
+
+  name: string;
+}
+
+class VariableDeclaration implements CodeGenerator {
```

Comment:
[Request] This code to generate tests is nifty, but really hard for me to read through. I'm starting to feel more and more negative about autogenerated tests over time - it's hard to identify what all the tests are, and the code tends to be moderately unreadable.

Oftentimes, code that's using generation like this for tests is creating a lot of redundant cases. I.e. if there are cases `A`, `B`, `C`, and `D`, they create `A * B * C * D`. But there often don't need to be so many tests for all possible permutations! Just individual tests for each of them (`A + B + C + D`).

Could you please change this file to have string literals for the tests? If there's a reason that's not practical we can discuss, but my guess for this small a rule we don't need so much testing complication.